### ISSUE TITLE
imbalanced classes

### ISSUE THREAD

**Issue by emoebel** [eriklindernoren/PyTorch-YOLOv3#738]
Hi, I would like to try this yolo implementation on my dataset, which has imbalanced classes. How can I address this issue with PyTorch-YOLOv3 ? I don't see any "class-weights" or "re-sampling" options

**Comment by Flova** [comment]
Hi,
this hyperparameter is not implemented, as we are mostly following the `.cfg` of the original darknet implementation which does not include this param afaik. Also, I want to keep the number of hyperparameters down a bit to simplify things. But I get that this may be a concern for very unbalanced classes. I mostly only read and tried things (e.g. focal loss) to help with the object / no object classification, which is heavily imbalanced to the no obj side in the YOLO family, but the imbalance of object classes is (most of the time) multiple magnitudes lower and not so much of a concern for my usecases. How large is your imbalance? 
A quick and dirty fix would be to add balancing values in line 69 in the `loss.py` file to the definition of the `BCELoss`. You can pass an array with a weight for each of your classes to the constructor parameter `pos_weight`.

**Comment by emoebel** [comment]
Number of instances per class: class1=5000, class2=300
Without taking into account this imbalance, Yolo does not find any of the class2 objects. I have already tried to use class_weights for Mask-RCNN and Stardist implementations, which did improve detection of class2.

I am not used to Pytorch, I come from the Keras framework. So how exactly should I modify the line you mention in the `loss.py` file ?
```
BCEcls = nn.BCEWithLogitsLoss(
        pos_weight=torch.tensor([1.0], device=device))
```

**Comment by Flova** [comment]
I don't know if rescaling the classification helps if the object is not detected at all. Because yolo has a separated loss for object detection and classification. The part that detects an object doesn't care about its class, so it might be an issue where the sample size is too small regardless of the other class.

But if you want to try it using classification loss weights, you need to modify the `pos_weight` argument in the said line to `pos_weight=torch.tensor([weight1, weight2], device=device)` where weight1 and weight 2 are the weights for class 1 and 2. You need to add a weight for all n classes. The values itself are hyperparameters and need to be determined for your data distribution.
